ReactNativeClient/lib/services/KeymapService.ts

```diff
@@ -0,0 +1,284 @@
+		} catch (err) {
+			this.resetKeymap();
+			throw new Error(
+				`Keymap configuration contains duplicates!\n${err.message}`
+			);
+		}
+	}
+
+	resetKeymap() {
+		this.keymap = KeymapService.initKeymap();
+	}
+
+	private validateKeymapItem(item: KeymapItem) {
+		const commandReference = 'https://github.com/laurent22/joplin/blob/master/README.md';
+		const acceleratorReference = 'https://www.electronjs.org/docs/api/accelerator';
+
+		if (!item.hasOwnProperty('command')) {
+			throw new Error(
+				'Couldn\'t find the "command" property.\n\n' +
+				`Visit ${commandReference} to see the list of available commands.`
+			);
+		} else if (!this.keymap.hasOwnProperty(item.command)) {
+			throw new Error(
+				`"${item.command}" is not a valid command.\n\n` +
+				`Visit ${commandReference} to see the list of available commands.`
+			);
+		}
+
+		if (!item.hasOwnProperty('accelerator')) {
+			throw new Error(
+				'Couldn\'t find the "accelerator" property.\n\n' +
+				`Visit ${acceleratorReference} to learn more about Electron Accelerators.`
+			);
+		} else if (!(item.accelerator === null || KeymapService.isAccelerator(item.accelerator))) {
+			throw new Error(
+				`"${item.accelerator}" is not a valid Accelerator.\n\n` +
+				`Visit ${acceleratorReference} to learn more about Electron Accelerators.`
+			);
+		}
+	}
+
+	private validateKeymap() {
+		const usedAccelerators = new Set();
+
+		for (const item of Object.values(this.keymap)) {
+			const itemAccelerator = item.accelerator;
+
+			if (usedAccelerators.has(itemAccelerator)) {
+				const originalItem = Object.values(this.keymap).find(_item => _item.accelerator == item.accelerator);
+
+				throw new Error(
+					`Accelerator "${itemAccelerator}" can't be used for "${item.command}" command ` +
+					`because it's already used for "${originalItem.command}" command!\n\n` +
+					'You have to change the Accelerator for any of above commands.'
+				);
+			} else if (itemAccelerator !== null) {
+				usedAccelerators.add(itemAccelerator);
+			}
+		}
+	}
+
+	static initKeymap() {
+		const itemsByCommand: { [command: string]: KeymapItem } = {};
+		for (let i = 0; i < KeymapService.defaultKeymap.length; i++) {
+			// Make a copy of the KeymapItem before assigning it, hence the spread operator
+			// Otherwise we're going to mess up the defaultKeymap array
+			itemsByCommand[KeymapService.defaultKeymap[i].command] = { ...KeymapService.defaultKeymap[i] };
+		}
+
+		return itemsByCommand;
+	}
+
+	static isAccelerator(accelerator: string) {
+		let keyFound = false;
+		const parts = accelerator.split('+');
+
+		return parts.every((part, index) => {
+			const isKey = KeymapService.keyCodes.test(part);
+			const isModifier = KeymapService.modifiers.test(part);
+
+			if (isKey) {
+				// Key must be unique
+				if (keyFound) return false;
+				keyFound = true;
+			}
+
+			// Key is required
+			if (index === (parts.length - 1) && !keyFound) return false;
+			return isKey || isModifier;
+		});
+	}
```

Comment:
What does this function do?

Reply:
It's for validating accelerators. Electron silently disables shortcuts if we provide invalid accelerators for menu items. This function is used for validating accelerators in the `keymap-desktop.json` file. The program will crash if the user provides invalid accelerators.

Reply:
This currently accepts modifiers like CommandOrControl, CmdOrCtrl because I was planning to support all valid Electron accelerators. But if we are strictly not accepting those modifiers, I'll have to make some changes.

Reply:
`isAccelerator` is perhaps a bit vague - for example it could be to check if the provided accelerator is currently defined or not. Since the function is for validation, perhaps `validateAccelerator` would be better?